### ISSUE TITLE
Missing redirect on token cache miss?

### ISSUE THREAD

**Issue by HenningRoigaard** [Azure-Samples/active-directory-aspnetcore-webapp-openidconnect-v2#150]
> Please provide us with the following information:
> ---------------------------------------------------------------

### This issue is for a: (mark with an `x`)
```
- [x] bug report -> please search issues before submitting
- [ ] feature request
- [ ] documentation issue or request
- [ ] regression (a behavior that used to work and stopped in a new release)
```

### The issue was found for the following scenario:

Please add an 'x' for the scenario(s) where you found an issue

1. [ ] Web app that signs in users
   1. [ ] with a work and school account in your organization: [1-WebApp-OIDC/1-1-MyOrg](../blob/master/1-WebApp-OIDC/1-1-MyOrg)
   1. [ ] with any work and school account: [/1-WebApp-OIDC/1-2-AnyOrg](../blob/master/1-WebApp-OIDC/1-2-AnyOrg)
   1. [ ] with any work or school account or Microsoft personal account: [1-WebApp-OIDC/1-3-AnyOrgOrPersonal](../blob/master/1-WebApp-OIDC/1-3-AnyOrgOrPersonal)
   1. [ ] with users in National or sovereign clouds [1-WebApp-OIDC/1-4-Sovereign](../blob/master/1-WebApp-OIDC/1-4-Sovereign)
   1. [ ] with B2C users [1-WebApp-OIDC/1-5-B2C](../blob/master/1-WebApp-OIDC/1-5-B2C)
1. Web app that calls Microsoft Graph
   1. [ ] Calling graph with the Microsoft Graph SDK: [2-WebApp-graph-user/2-1-Call-MSGraph](../blob/master/2-WebApp-graph-user/2-1-Call-MSGraph)
   1. [ ] With specific token caches: [2-WebApp-graph-user/2-2-TokenCache](../blob/master/2-WebApp-graph-user/2-2-TokenCache)
   1. [ ] Calling Microsoft Graph in national clouds: [2-WebApp-graph-user/2-4-Sovereign-Call-MSGraph](../blob/master/2-WebApp-graph-user/2-4-Sovereign-Call-MSGraph)
1. [ ] Web app calling several APIs [3-WebApp-multi-APIs](../blob/master/3-WebApp-multi-APIs)
1. [x] Web app calling your own Web API
1. Web app restricting users
   1. [ ] by Roles: [5-WebApp-AuthZ/5-1-Roles](../blob/master/5-WebApp-AuthZ/5-1-Roles)
   1. [ ] by Groups: [5-WebApp-AuthZ/5-2-Groups](../blob/master/5-WebApp-AuthZ/5-2-Groups)
1. [ ] Deployment to Azure
1. [ ] Other (please describe)

I have observed a behavior, which I believe to be wrong. However, I am unsure what the correct behavior is, and where it should be implemented.

The basic issue is this: 

The normal auth flow involves adding the user’s tokens to the token cache in the client app, and giving the user’s browser an auth cookie (.AspNetCore.AzureADCookie). If, for whatever reason, the token is removed from the token cache in the client app, I would expect the user to be redirected to login on next request to the client app (or the next request using the token cache at least). However, if the browser request includes the auth cookie, the observed behavior is that the client app accepts the cookie, does not get new tokens, and continues, as if nothing was amiss. An error then occurs on the first attempt to acquire an access token, and the user receives an error code 500.

This behavior can be reproduced using the 4-WebApp-your-API sample as follows:

1.	Run the solution (F5 in VS). Browser opens automatically and redirects to login.
2.	Log-in and validate that the “TodoList” links works
3.	Flush the InMemoryTokenCache by closing the browser and the two services
a.	Ensure that cookies do not get deleted automatically on browser close
4.	Run the solution (F5 in VS). This time we are not redirected to login, nor do we silently acquire the missing tokens.
5.	Use the “TodoList” link and get an error code 500

The underlying exception is:

```
Microsoft.Identity.Client.MsalUiRequiredException: 
No account or login hint was passed to the AcquireTokenSilent call.
   at Microsoft.Identity.Client.AcquireTokenSilentParameterBuilder.Validate()
   at Microsoft.Identity.Client.AbstractAcquireTokenParameterBuilder`1.ValidateAndCalculateApiId()
   at Microsoft.Identity.Client.AbstractClientAppBaseAcquireTokenParameterBuilder`1.ExecuteAsync(CancellationToken cancellationToken)
   at Microsoft.Identity.Client.AbstractAcquireTokenParameterBuilder`1.ExecuteAsync()
   at Microsoft.Identity.Web.Client.TokenAcquisition.<GetAccessTokenOnBehalfOfUser>d__15.MoveNext() in F:\src\git\External\active-directory-aspnetcore-webapp-openidconnect-v2\Microsoft.Identity.Web\Client\TokenAcquisition.cs:line 381
```
If you delete the .AspNetCore.AzureADCookie before step (4) above, the behavior is different: The token cache is silently populated, and everything works as expected. I expect that the client app redirects as expected, but some cookies in the login.microsoft.com domain allows the browser to silently authenticate.

**Comment by TiagoBrenck** [comment]
Hi @HenningRoigaard, the PR https://github.com/Azure-Samples/active-directory-aspnetcore-webapp-openidconnect-v2/pull/149 is taking care of this. However this exception is actually by design and is up to the developers to decide what to do when it gets thrown.

The PR in review uses a filter that will try to re-login the user in case this exception is thrown, but this solution may not fit all the business cases out there. It depends on the user-case to decide what to do when this exception is thrown.

**Comment by HenningRoigaard** [comment]
Using MsalUiRequiredExceptionFilterAttribute solved the issue (The PR #149 isn't required for simple use cases)
